### PR TITLE
Items

### LINKED ISSUE #13: User Story 17, Items Index Page
As any kind of user on the system
I can visit the items catalog ("/items")
I see all items in the system except disabled items

The item image is a link to that item's show page

### LINKED ISSUE #16: User Story 18, Items Index Page Statistics
User Story 18, Items Index Page Statistics

As any kind of user on the system
When I visit the items index page ("/items")
I see an area with statistics:
- the top 5 most popular items by quantity purchased, plus the quantity bought
- the bottom 5 least popular items, plus the quantity bought

"Popularity" is determined by total quantity of that item ordered

### PR DESCRIPTION
**Pull request by hashmaster3k**
User story 17 and 18

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,7 +1,7 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
 
-  helper_method :cart, :current_user
+  helper_method :cart, :current_user, :current_merchant?, :current_admin?
```
Comment (hashmaster3k) [inline app/controllers/application_controller.rb:4]: Adds helper methods to allow nav bar to access methods

Comment (dcoleman21) [inline app/controllers/application_controller.rb:4]: Glad we got this sorted out

File: app/controllers/orders_controller.rb [hunk 2]
```diff
@@ -17,6 +17,7 @@ def create
           quantity: quantity,
           price: item.price
           })
+          item.update_quantity_purchased(quantity)
```
Comment (hashmaster3k) [inline app/controllers/orders_controller.rb:20]: Updates item purchased quantity with every order

File: app/views/items/index.html.erb [hunk 3]
```diff
@@ -4,22 +4,34 @@
...
+    <% end %>
+  </ol>
+</section>
 <section class="grid-container">
   <% @items.each do |item| %>
     <section class = "grid-item" id= 'item-<%=item.id%>'>
       <h2> <%=link_to item.name, "/items/#{item.id}" %> </h2>
       <p>Sold by: <%=link_to item.merchant.name, "/merchants/#{item.merchant.id}" %></p>
-      <img src= <%= item.image %>>
+      <%= link_to image_tag(item.image), "/items/#{item.id}" %>
```
Comment (dcoleman21) [inline app/views/items/index.html.erb:29]: Created the image as a link

File: app/models/item.rb [hunk 4]
```diff
@@ -25,4 +25,21 @@ def no_orders?
...
+    order(quantity_purchased: :asc).limit(5)
+  end
+
+  def self.active_items
+    self.where(active?: true)
+  end
+
+  def update_quantity_purchased(quantity)
+    self.quantity_purchased += quantity
+    self.save
```
Comment (hashmaster3k) [inline app/models/item.rb:42]: Unsure if there is a better way to do this.

File: spec/features/items/index_spec.rb [hunk 5]
```diff
@@ -6,10 +6,11 @@
...
 
-      @tire = @meg.items.create(name: "Gatorskins", description: "They'll never pop!", price: 100, image: "https://www.rei.com/media/4e1f5b05-27ef-4267-bb9a-14e35935f218?size=784x588", inventory: 12)
+      @tire = @meg.items.create!(name: "Gatorskins", description: "They'll never pop!", price: 100, image: "https://www.rei.com/media/4e1f5b05-27ef-4267-bb9a-14e35935f218?size=784x588", inventory: 12, quantity_purchased: 2)
 
-      @pull_toy = @brian.items.create(name: "Pull Toy", description: "Great pull toy!", price: 10, image: "http://lovencaretoys.com/image/cache/dog/tug-toy-dog-pull-9010_2-800x800.jpg", inventory: 32)
-      @dog_bone = @brian.items.create(name: "Dog Bone", description: "They'll love it!", price: 21, image: "https://img.chewy.com/is/image/catalog/54226_MAIN._AC_SL1500_V1534449573_.jpg", active?:false, inventory: 21)
+      @pull_toy = @brian.items.create!(name: "Pull Toy", description: "Great pull toy!", price: 10, image: "http://lovencaretoys.com/image/cache/dog/tug-toy-dog-pull-9010_2-800x800.jpg", inventory: 32, quantity_purchased: 3)
```
Comment (dcoleman21) [inline spec/features/items/index_spec.rb:11]: Added quantity purchased here as well

File: spec/features/orders/creation_spec.rb [hunk 6]
```diff
@@ -1,19 +1,13 @@
-# When I fill out all information on the new order page
-# And click on 'Create Order'
-# An order is created and saved in the database
-# And I am redirected to that order's show page with the following information:
-#
-# - Details of the order:
-
-# - the date when the order was created
+require 'rails_helper'
```
Comment (dcoleman21) [inline spec/features/orders/creation_spec.rb:1]: Don't forget to put in 'rails_helper' lol

**Review by dcoleman21 — APPROVED** [review]

File: db/schema.rb [hunk 7]
```diff
@@ -36,6 +36,7 @@
     t.bigint "merchant_id"
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
+    t.integer "quantity_purchased", default: 0, null: false
```
Comment (hashmaster3k) [inline db/schema.rb:39]: Was having a strange issue where in development quantity was defaulting to 0 but the test environment was nil

**Comment by fieldstyler** [comment]
Reviewed by: Tyler, Dani

**Review by fieldstyler — APPROVED** [review]

**Pull request merged**